File: VB/SnapModule/Controllers/InplaceDocumentsController.vb
Imports System
Imports System.Collections.Generic
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports SnapModule.Base
Imports DevExpress.Persistent.Base
Imports DevExpress.Data.Filtering
Imports System.Collections
Imports SnapModule.Helpers

Namespace SnapModule.Controllers
    Public Class InplaceDocumentsController
        Inherits ObjectViewController

        Private documentPrintSelectionAction As SingleChoiceAction
        Private inplaceDocumentsCacheHelper As InplaceDocumentsCacheHelper
        Public Sub New()
            documentPrintSelectionAction = New SingleChoiceAction(Me, "DocumentPrintSelection", PredefinedCategory.Reports)
            documentPrintSelectionAction.Caption = "Print Selected Objects"
            AddHandler documentPrintSelectionAction.Execute, AddressOf documentPrintSelectionAction_Execute
            documentPrintSelectionAction.ItemType = SingleChoiceActionItemType.ItemIsOperation
            documentPrintSelectionAction.SelectionDependencyType = SelectionDependencyType.RequireMultipleObjects
            documentPrintSelectionAction.ImageName = "Action_Printing_Preview"
        End Sub
        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            Dim documentDataType As Type = Application.Modules.FindModule(Of SnapModule)().DocumentDataType
            inplaceDocumentsCacheHelper = New InplaceDocumentsCacheHelper(Application.CreateObjectSpace(documentDataType), documentDataType)
            Dim documents As IList(Of IDocumentData) = inplaceDocumentsCacheHelper.GetDocumentDataList(View.ObjectTypeInfo.Type)
            Dim items As New List(Of ChoiceActionItem)()
            For Each document As IDocumentData In documents
                Dim item As New ChoiceActionItem(document.DocumentName, document)
                items.Add(item)
            Next document
            items.Sort(Function(left As ChoiceActionItem, right As ChoiceActionItem) Comparer(Of String).Default.Compare(left.Caption, right.Caption))
            documentPrintSelectionAction.Items.Clear()
            documentPrintSelectionAction.Items.AddRange(items)
        End Sub
        Protected Overrides Sub OnDeactivated()
            MyBase.OnDeactivated()
            inplaceDocumentsCacheHelper.ClearInplaceDocumentsCache()
            inplaceDocumentsCacheHelper = Nothing
        End Sub
        Protected Overrides Sub OnFrameAssigned()
            MyBase.OnFrameAssigned()
            If (Frame IsNot Nothing) AndAlso ((Frame.Context = TemplateContext.LookupWindow) OrElse (Frame.Context = TemplateContext.LookupControl)) Then
                Me.Active.SetItemValue("DisableForLookupListView", False)
            End If
        End Sub
        Private Sub documentPrintSelectionAction_Execute(ByVal sender As Object, ByVal e As SingleChoiceActionExecuteEventArgs)
            If TypeOf e.SelectedChoiceActionItem.Data Is IDocumentData Then
                Dim criteria As CriteriaOperator = CType(ObjectSpace, BaseObjectSpace).GetObjectsCriteria(View.ObjectTypeInfo, e.SelectedObjects)
                SnapDocumentHelper.ShowPrintPreview(DirectCast(e.SelectedChoiceActionItem.Data, IDocumentData), Application.CreateObjectSpace(View.ObjectTypeInfo.Type), criteria)
            End If
        End Sub
    End Class
End Namespace

File: CS/SnapModule/Controllers/InplaceDocumentsController.cs
using System;
using System.Collections.Generic;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using SnapModule.Base;
using DevExpress.Persistent.Base;
using DevExpress.Data.Filtering;
using System.Collections;
using SnapModule.Helpers;

namespace SnapModule.Controllers {
    public class InplaceDocumentsController : ObjectViewController {
        private SingleChoiceAction documentPrintSelectionAction;
        private InplaceDocumentsCacheHelper inplaceDocumentsCacheHelper;
        public InplaceDocumentsController() {
            documentPrintSelectionAction = new SingleChoiceAction(this, "DocumentPrintSelection", PredefinedCategory.Reports);
            documentPrintSelectionAction.Caption = "Print Selected Objects";
            documentPrintSelectionAction.Execute += new SingleChoiceActionExecuteEventHandler(documentPrintSelectionAction_Execute);
            documentPrintSelectionAction.ItemType = SingleChoiceActionItemType.ItemIsOperation;
            documentPrintSelectionAction.SelectionDependencyType = SelectionDependencyType.RequireMultipleObjects;
            documentPrintSelectionAction.ImageName = "Action_Printing_Preview";
        }
        protected override void OnActivated() {
            base.OnActivated();
            Type documentDataType = Application.Modules.FindModule<SnapModule>().DocumentDataType;
            inplaceDocumentsCacheHelper = new InplaceDocumentsCacheHelper(Application.CreateObjectSpace(documentDataType), documentDataType);
            IList<IDocumentData> documents = inplaceDocumentsCacheHelper.GetDocumentDataList(View.ObjectTypeInfo.Type);
            List<ChoiceActionItem> items = new List<ChoiceActionItem>();
            foreach (IDocumentData document in documents) {
                ChoiceActionItem item = new ChoiceActionItem(document.DocumentName, document);
                items.Add(item);
            }
            items.Sort(delegate(ChoiceActionItem left, ChoiceActionItem right) {
                return Comparer<string>.Default.Compare(left.Caption, right.Caption);
            });
            documentPrintSelectionAction.Items.Clear();
            documentPrintSelectionAction.Items.AddRange(items);
        }
        protected override void OnDeactivated() {
            base.OnDeactivated();
            inplaceDocumentsCacheHelper.ClearInplaceDocumentsCache();
            inplaceDocumentsCacheHelper = null;
        }
        protected override void OnFrameAssigned() {
            base.OnFrameAssigned();
            if ((Frame != null) && ((Frame.Context == TemplateContext.LookupWindow) || (Frame.Context == TemplateContext.LookupControl))) {
                this.Active.SetItemValue("DisableForLookupListView", false);
            }
        }
        private void documentPrintSelectionAction_Execute(object sender, SingleChoiceActionExecuteEventArgs e) {
            if (e.SelectedChoiceActionItem.Data is IDocumentData) {
                CriteriaOperator criteria = ((BaseObjectSpace)ObjectSpace).GetObjectsCriteria(View.ObjectTypeInfo, e.SelectedObjects);
                SnapDocumentHelper.ShowPrintPreview((IDocumentData)e.SelectedChoiceActionItem.Data, Application.CreateObjectSpace(View.ObjectTypeInfo.Type), criteria);
            }
        }
    }
}
